### PR TITLE
Create python REPL for debugging drakvuf

### PR DESCRIPTION
**Pull request by BonusPlay**
This PR adds easily embeddable IPython REPL for debugging drakvuf hooks.

Running: `ctypesgen src/libdrakvuf/libdrakvuf.h -o src/librepl/libdrakvuf.py -I /usr/include -I /usr/include/glib-2.0 -I /lib/x86_64-linux-gnu/glib-2.0/include` generated ctypes bindings for python, which are later imported by REPL. Would be nice to incorporate this command into build process.

This PR also includes a `repl` binary which runs standalone drakvuf and starts REPL on first CR3 breakpoint (to try it out).

Currently there is no way to modify vmi return code, so editing registers is not implemented, but this is a subject to change.

image

### CODE REVIEW

**Comment by icedevml** [comment]
@drakvuf-jenkins This is OK to test.

File: src/repl.cpp [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+static event_response_t trap_handler(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
+{
+	auto ret = repl_start(drakvuf, info);
+	drakvuf_remove_trap(drakvuf, info->trap, (drakvuf_trap_free_t)free);
+	return ret;
+}
+
+int main(int argc, char** argv)
+{
+	int return_code = 0;
```
Comment (icedevml) [inline src/repl.cpp]: tabs vs spaces in indentation?

**Comment by icedevml** [comment]
@BonusPlay 

```
configure.ac:107: installing './missing'
src/Makefile.am: installing './depcomp'
src/librepl/Makefile.am:148: error: librepl_la_LIBADD must be set with '=' before using '+='
autoreconf: automake failed with exit status: 1
Build step 'Execute shell' marked build as failure
Skipped archiving because build is not successful
Finished: FAILURE
```

**Comment by BonusPlay** [comment]
Yes, this was a mistake while rebasing.

**Comment by icedevml** [comment]
```
config.status: creating src/dirwatch/Makefile
config.status: creating src/libusermode/Makefile
config.status: creating src/librepl/Makefile
config.status: creating config.h
config.status: executing depfiles commands
config.status: error: in `/shared/jenkins/DRAKVUF-build':
config.status: error: Something went wrong bootstrapping makefile fragments
    for automatic dependency tracking.  Try re-running configure with the
    '--disable-dependency-tracking' option to at least be able to build
    the package (albeit without support for automatic dependency tracking).
See `config.log' for more details
Build step 'Execute shell' marked build as failure
Skipped archiving because build is not successful
Finished: FAILURE
```

:thinking:

**Comment by BonusPlay** [comment]
> `Build step 'Execute shell' marked build as failure`

Sounds like `execute shell` is not allowed, however this is a the proper way to link python libs to project.

> `Making all in librepl
make[3]: Entering directory '/home/travis/build/tklengyel/drakvuf/src/librepl'
Makefile:649: *** missing separator.  Stop.`

Hmm, it does work on my development machine...

File: configure.ac [hunk 2]
```diff
@@ -545,6 +545,7 @@ AC_CONFIG_FILES([Makefile src/Makefile
                  src/plugins/Makefile
                  src/dirwatch/Makefile
                  src/libusermode/Makefile
+                 src/librepl/Makefile
```
Comment (tklengyel) [inline configure.ac]: This new library should be conditionally enabled (`--enable-repl`) and be off by default.

**Comment by icedevml** [comment]
@tklengyel In terms of this PR, this is some attempt to have (at least very basic) live debugging possibilities with DRAKVUF. As @BonusPlay wrote in the PR description, it adds an alternative entry point in a form o `repl` binary. Inside the Python REPL, it is possible to add some new traps and then to inspect `drakvuf_trap_info_t` by hand, when the callback hits.

**Comment by tklengyel** [comment]
@icedevml that's all fine and good, my only concerns is `drakvuf_LDADD += librepl/librepl.la` adding this new library for all builds, even for ones where it's not needed

**Comment by icedevml** [comment]
@tklengyel AFAIK this is already fixed:

https://github.com/tklengyel/drakvuf/pull/824/files#diff-480477e89f9b6ddafb30c4383dcdd705R162-R174

File: src/librepl/Makefile.am [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+sources =
+sources += librepl.h
+sources += repl.cpp
+
+AM_CPPFLAGS = -I$(top_srcdir) -I$(top_srcdir)/src -I$(srcdir)
+AM_CPPFLAGS += $(CPPFLAGS)
+AM_CPPFLAGS += $(VMI_CFLAGS)
+AM_CPPFLAGS += $(GLIB_CFLAGS)
+AM_CPPFLAGS += $(JSONC_CFLAGS)
+AM_CPPFLAGS += $(shell python3-config --cflags)
```
Comment (tklengyel) [inline src/librepl/Makefile.am]: We should probably do this via configure.ac similar to https://github.com/nasa/trick/blob/master/autoconf/configure.ac#L363

File: src/librepl/Makefile.am [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+
+AM_LDFLAGS =  $(LDFLAGS)
+AM_LDFLAGS += $(VMI_LIBS)
+AM_LDFLAGS += $(GLIB_LIBS)
+AM_LDFLAGS += $(JSONC_LIBS)
+AM_LDFLAGS += $(shell python3-config --ldflags)
```
Comment (tklengyel) [inline src/librepl/Makefile.am]: Same as above

**Comment by BonusPlay** [comment]
@tklengyel that should have done it. I'm generating `libdrakvuf.py` by doing
```
ctypesgen src/libdrakvuf/libdrakvuf.h -o src/librepl/libdrakvuf.py -I /usr/include -I /usr/include/glib-2.0 -I /lib/x86_64-linux-gnu/glib-2.0/include
```
Where `ctypesgen` is a python library. I don't know autotools well enough to implement this command into the build system myself. I should mention somewhere that IPython needs to be installed as well.

**Comment by tklengyel** [comment]
You should just google around for `configure.ac ctypesgen` and you will find some project that already figured it out. For example: https://github.com/p4lang/switch/blob/master/configure.ac https://github.com/p4lang/switch/blob/92afb048c2e7f8691e01b4e3839bd7acfd3de4b5/switchapi/Makefile.am#L31

File: configure.ac [hunk 5]
```diff
@@ -116,6 +116,30 @@ AX_CXX_COMPILE_STDCXX_11
 
 ######################################################
 
+AC_ARG_ENABLE([repl],
+  [AS_HELP_STRING([--enable-repl],
+    [Enable the repl library @<:@no@:>@])],
+  [repl="$enableval"],
+  [repl="no"])
+AM_CONDITIONAL([REPL], [test x$repl = xyes])
+if test x$repl = xyes; then
```
Comment (tklengyel) [inline configure.ac]: This part (within the if test ...) needs to be moved further down in configure.ac, after `AC_CHECK_LIB([xentoollog], [xtl_logger_destroy])`

**Comment by tklengyel** [comment]
@drakvuf-jenkins Retest this please

File: src/librepl/repl.cpp [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+    Py_DECREF(refCount);
+
+static void repl_init(drakvuf_t drakvuf)
+{
+    // init python
+    Py_Initialize();
+
+    // load libdrakvuf
+    PyObject* sysPath = PySys_GetObject("path");
+    PyList_Append(sysPath, PyUnicode_FromString("/opt/drakvuf/src/librepl"));
```
Comment (BonusPlay) [inline src/librepl/repl.cpp]: imo we should change this string, but the problem is that python needs absolute path here

Comment (tklengyel) [reply src/librepl/repl.cpp]: Make it configurable somehow?

**Comment by BonusPlay** [comment]
@tklengyel would be nice to add it to at least 1 task on travis (since currently all builds omit librepl), what do you think?

**Comment by BonusPlay** [comment]
hmmm, exactly what I was talking about...

**Comment by tklengyel** [comment]
Adding a travis task sounds good to me, would be very appropriate

**Comment by BonusPlay** [comment]
I've added just debug build (as 99.99% of the time it will/should be used in that build) to not add too much stress to CI/CD.

**Comment by BonusPlay** [comment]
Did take quite a few attempts, but hey - it works :)

File: configure.ac [hunk 7]
```diff
@@ -557,6 +587,8 @@ Host system type: $host
 Build system type: $build
 Installation prefix: $prefix
 -------------------------------------------------------------------------------
+REPL:         $repl
```
Comment (tklengyel) [inline configure.ac]: Could you place this after the plugins?

Comment (BonusPlay) [reply configure.ac]: Done!

**Pull request merged**